### PR TITLE
Models docs: Add examples for custom python models and improve custom flavors documentation

### PR DESCRIPTION
**Pull request by dbczumar**
This PR adds examples that walk through the process of creating custom Python models using `mlflow.pyfunc.save/log_model`. It also provides greater clarity about writing custom model flavors and about the `python_function` flavor.

### CODE REVIEW

File: docs/source/models.rst [hunk 1]
```diff
@@ -182,11 +189,172 @@ TensorFlow (``tensorflow``)
...
+dependencies. These artifact dependencies may include serialized models produced by any Python ML 
+library.
+
+Because these custom models contain the ``python_function`` flavor, they can be deployed
+to any of MLflow's supported production environments, such as SageMaker, AzureML, or local
+REST endpoints.
+
+The following examples demonstrate how the :py:mod:`mlflow.pyfunc` module can be used to create 
+custom Python models. For additional information about model customization with MLflow's 
+``python_function`` utilities, see the
```
Comment (stbof) [inline docs/source/models.rst]: The following examples demonstrate how the :py:mod:`mlflow.pyfunc` module can be used to create 
custom Python models. -> The following examples demonstrate how you can use the :py:mod:`mlflow.pyfunc` module to create custom Python models.

File: docs/source/models.rst [hunk 2]
```diff
@@ -182,11 +189,172 @@ TensorFlow (``tensorflow``)
...
+
+The following examples demonstrate how the :py:mod:`mlflow.pyfunc` module can be used to create 
+custom Python models. For additional information about model customization with MLflow's 
+``python_function`` utilities, see the 
+:ref:`python_function custom models documentation <pyfunc-create-custom>`.
+
+Example: Creating a custom "add n" model
+****************************************
+In this example, we first define a class for a custom model that adds a specified numeric value, 
+`n`, to all columns of a Pandas DataFrame input. Then, we leverage the :py:mod:`mlflow.pyfunc` APIs
```
Comment (stbof) [inline docs/source/models.rst]: In this example, we first define a class -> This example defines a class 

Then, we leverage the :py:mod:`mlflow.pyfunc` APIs -> Then, it uses the :py:mod:`mlflow.pyfunc` APIs 

Finally, we load the model in  ``python_function`` format and use it to evaluate a sample input -> Finally, it loads the model in  ``python_function`` format and uses it to evaluate a sample input

File: docs/source/models.rst [hunk 3]
```diff
@@ -182,11 +189,172 @@ TensorFlow (``tensorflow``)
...
+    import pandas as pd
+    model_input = pd.DataFrame([range(10)])
+    model_output = loaded_model.predict(model_input)
+    assert model_output.equals(pd.DataFrame([range(5, 15)]))
+
+
+Example: Saving an XGBoost model in MLflow format
+*************************************************
+In this example, we begin by training and saving a gradient boosted tree model using the XGBoost
+library. Next, we define a wrapper class around the XGBoost model that conforms to MLflow's
```
Comment (stbof) [inline docs/source/models.rst]: In this example, we begin by training and saving a gradient boosted tree model using the XGBoost
library.  -> This example begins by training and saving a gradient boosted tree model using the XGBoost
library. 

Next, we define a wrapper class -> Next, it defines a wrapper class

Finally, we load the MLflow model  ... and use it -> Finally, it loads the MLflow model ... and uses it

File: docs/source/models.rst [hunk 4]
```diff
@@ -182,11 +189,172 @@ TensorFlow (``tensorflow``)
...
+^^^^^^^^^^^^^^
+You can also create custom MLflow models by writing a custom *flavor*.
+
+As discussed in the :ref:`model-api` and :ref:`model-storage-format` sections, an MLflow model
+is defined by a directory of files that contains an ``MLmodel`` configuration file. This ``MLmodel``
+file describes various model attributes, including the flavors in which the model can be 
+interpreted. The ``MLmodel`` file contains an entry for each flavor name; each entry is
+a YAML-formatted collection of flavor-specific attributes.
+
+In order to create a new flavor to support a custom model, you simply need to define the set
```
Comment (stbof) [inline docs/source/models.rst]: In order to create a new flavor to support a custom model, you simply need to define the set  -> To create a new flavor to support a custom model, you define the set

File: docs/source/models.rst [hunk 5]
```diff
@@ -182,11 +189,172 @@ TensorFlow (``tensorflow``)
...
+of flavor-specific attributes to include in the ``MLmodel`` configuration file, as well as the
+code that can interpret the contents of the model directory and the flavor's attributes.
+
+As an example, let's examine the :py:mod:`mlflow.pytorch` module corresponding to MLflow's 
+``pytorch`` flavor. In the :py:func:`mlflow.pytorch.save_model()` method, a PyTorch model is saved 
+to a specified output directory. Additionally, :py:func:`mlflow.pytorch.save_model()` leverages the 
+:py:func:`mlflow.models.Model.add_flavor()` and :py:func:`mlflow.models.Model.save()` functions to 
+produce an ``MLmodel`` configuration containing the ``pytorch`` flavor. The resulting configuration 
+has several flavor-specific attributes, such as ``pytorch_version``, which denotes the version of the 
+PyTorch library that was used to train the model. In order to interpret model directories
```
Comment (stbof) [inline docs/source/models.rst]: In order to interpret model directories -> To interpret model directories

**Review by ghost — CHANGES_REQUESTED** [review]
Main comments focus on removing extra words, and sticking to second person "you", rather than introducing third person plural "we".

File: docs/source/models.rst (hunk 1 above)
Comment (dbczumar) [reply docs/source/models.rst]: Changed to active voice. Thanks!

File: docs/source/models.rst (hunk 2 above)
Comment (dbczumar) [reply docs/source/models.rst]: Made these fixes! Thanks for providing the corrected text!

File: docs/source/models.rst (hunk 3 above)
Comment (dbczumar) [reply docs/source/models.rst]: Fixed!

File: docs/source/models.rst (hunk 4 above)
Comment (dbczumar) [reply docs/source/models.rst]: Fixed! Thanks

File: docs/source/models.rst (hunk 5 above)
Comment (dbczumar) [reply docs/source/models.rst]: Done!

**Review by dbczumar — COMMENTED** [review]
@stbof Thanks for the review! Addressed your comments!

File: docs/source/models.rst [hunk 6]
```diff
@@ -182,11 +189,171 @@ TensorFlow (``tensorflow``)
...
+            input_matrix = xgb.DMatrix(model_input.values)
+            return self.xgb_model.predict(input_matrix)
+
+    # Create a Conda environment for the new MLflow model that contains the XGBoost library
+    # as a dependency, as well as the required CloudPickle library
+    import cloudpickle
+    conda_env = {
+        'channels': ['defaults'],
+        'dependencies': [
+          'xgboost={}'.format(xgb.__version__)
```
Comment (mparkhe) [inline docs/source/models.rst]: Missing a "comma" here

File: docs/source/models.rst [hunk 7]
```diff
@@ -182,11 +189,171 @@ TensorFlow (``tensorflow``)
 
 The ``tensorflow`` model flavor enables logging TensorFlow ``Saved Models`` and loading them back as ``Python Function`` models for inference on pandas DataFrames. Given a directory containing a saved model, you can log the model to MLflow via ``log_saved_model`` and then load the saved model for inference using ``mlflow.pyfunc.load_pyfunc``. For more information, see :py:mod:`mlflow.tensorflow`.
 
+Model Customization
+-------------------
+
+While MLflow's built-in model persistence utilities are convenient for packaging models from various
+popular ML libraries in MLflow model format, they do not cover every use case. For example, you may
+want to use a model from an ML library that is not explicitly supported by MLflow's built-in
+flavors. Alternatively, you may want to package arbitrary inference code and data to create an
```
Comment (mparkhe) [inline docs/source/models.rst]: Can we rephrase this ... "custom code" sounds more friendly than "arbitrary"

Comment (dbczumar) [reply docs/source/models.rst]: Done! Thanks

File: docs/source/models.rst (hunk 6 above)
Comment (dbczumar) [reply docs/source/models.rst]: Good catch! Thanks. Ran this example again to confirm that it works.

**Review by dbczumar — COMMENTED** [review]
@mparkhe Addressed your comments!

File: docs/source/models.rst [hunk 8]
```diff
@@ -134,7 +138,10 @@ Other MLflow components expect ``python_function`` models to follow this convent
 The ``python_function`` :ref:`model format <pyfunc-filesystem-format>` is defined as a directory
 structure containing all required data, code, and configuration.
 
-For more information, see the :mod:`mlflow.pyfunc` documentation.
+The :py:mod:`mlflow.pyfunc` module defines functions for saving and loading MLflows with the
+``python_function`` flavor. This module also includes utilities for creating custom Python models.
```
Comment (stbof) [inline docs/source/models.rst]: MLflows -> MLflow models

File: docs/source/models.rst [hunk 9]
```diff
@@ -182,12 +189,169 @@ TensorFlow (``tensorflow``)
...
+
+.. _custom-python-models:
+
+Custom Python Models
+^^^^^^^^^^^^^^^^^^^^
+The :py:mod:`mlflow.pyfunc` module provides :py:func:`save_model() <mlflow.pyfunc.save_model>` and
+:py:func:`log_model() <mlflow.pyfunc.log_model>` utilities for creating MLflow models with the
+``python_function`` flavor that contain  user-specified code and *artifact* (file) dependencies.
+These artifact dependencies may include serialized models produced by any Python MLlibrary.
+
```
Comment (stbof) [inline docs/source/models.rst]: any Python MLlibrary -> any Python ML library.

File: docs/source/models.rst (hunk 8 above)
Comment (dbczumar) [reply docs/source/models.rst]: Fixed!

File: docs/source/models.rst (hunk 9 above)
Comment (dbczumar) [reply docs/source/models.rst]: Fixed!

**Review by ghost — APPROVED** [review]

**Review by mparkhe — APPROVED** [review]
LGTM

**Pull request merged**